core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java

```diff
@@ -75,14 +86,105 @@ public DnsResolverGroupBuilder resolvedAddressTypes(
+     *         return 0; // Don't need to refresh a sporadically used domain.
+     *     }
+     *     ...
+     * });
+     * }</pre>
+     *
+     * <p>Note this method is mutually exclusive with {@link #autoRefreshTimeout(Duration)} and
+     * {@link #autoRefreshTimeoutMillis(long)}.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeout(ToLongFunction<? super String> timeoutFunction) {
+        requireNonNull(timeoutFunction, "timeoutFunction");
+        checkState(autoRefreshTimeoutFunction == null, "'autoRefreshTimeout()' was set already: %s",
+                   autoRefreshTimeoutFunction);
+        //noinspection unchecked
+        autoRefreshTimeoutFunction = (ToLongFunction<String>) timeoutFunction;
+        return this;
+    }
+
+    /**
+     * Sets the timeout after which a refreshing {@link DnsRecord} should expire.
+     * If this option is unspecified and {@link #enableAutoRefresh(boolean)} is set to
+     * {@code true}, a cached {@link DnsRecord} is automatically refreshed until the {@link ClientFactory}
+     * is closed.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeout(Duration timeout) {
+        requireNonNull(timeout, "timeout");
+        return autoRefreshTimeoutMillis(timeout.toMillis());
+    }
+
+    /**
+     * Sets the timeout in milliseconds after which a refreshing {@link DnsRecord} should expire.
+     * If this option is unspecified and {@link #enableAutoRefresh(boolean)} is set to
+     * {@code true}, a cached {@link DnsRecord} is automatically refreshed until the {@link ClientFactory}
+     * is closed.
+     */
+    @UnstableApi
+    public DnsResolverGroupBuilder autoRefreshTimeoutMillis(long timeoutMillis) {
+        checkArgument(timeoutMillis > 0, "timeoutMillis: %s (expected: > 0)", timeoutMillis);
+        return autoRefreshTimeout(hostname -> timeoutMillis);
+    }
+
     @Nullable
     MeterRegistry meterRegistry0() {
         return meterRegistry();
     }
 
     RefreshingAddressResolverGroup build(EventLoopGroup eventLoopGroup) {
-        return new RefreshingAddressResolverGroup(cacheSpec(), minTtl(), maxTtl(), negativeTtl(),
-                                                  refreshBackoff, resolvedAddressTypes, maybeCreateDnsCache(),
+        ToLongFunction<String> autoRefreshTimeoutFunction = this.autoRefreshTimeoutFunction;
+        Backoff autoRefreshBackoff = this.autoRefreshBackoff;
+        if (!autoRefresh && (autoRefreshTimeoutFunction != null || autoRefreshBackoff != null)) {
+            throw new IllegalStateException("Can't set 'autoRefreshTimeoutFunction' or 'autoRefreshBackoff' " +
+                                            "if 'autoRefresh' is disabled");
+        }
+
+        if (autoRefreshTimeoutFunction == null) {
+            autoRefreshTimeoutFunction = DEFAULT_AUTO_REFRESH_TIMEOUT_FUNCTION;
+        }
+        if (autoRefreshBackoff == null) {
+            autoRefreshBackoff = Backoff.ofDefault();
```

Comment:
I think it doesn't make sense to assign the `Backoff` to `autoRefreshBackoff` even when `autoRefresh` is `false`.
How about making `autoRefreshBackoff` as nullable and not passing `autoRefresh`?